### ISSUE TITLE
Fix false positives for functional pseudo-classes in no-descending-specificity

### ISSUE THREAD

**Issue by vegawong** [stylelint/stylelint#4010]
> Clearly describe the bug

There are some error tips in my selectors was not expected..
It flagged an error about rule ` no-descending-specificity`  when i use css-module presudo function like `:global()` with a selector.
Thought the selectors had not any relations ,it always flagged an error.
If i rewrite the same selectors without `:global`, then it works.

> Which rule, if any, is the bug related to?

no-descending-specificity

> What CSS is needed to reproduce the bug?

```css
.componentName :global(.ant-modal) :global(.ant-btn) {
    color: #fff;
}
.componentName :global(.ant-menu) {
    color: #fff;
}
```

> What stylelint configuration is needed to reproduce the bug?

e.g.

```javascript
module.exports = {
    extends: ['stylelint-config-standard', 'stylelint-config-css-modules', 'stylelint-config-prettier'],
    rules: {
        'at-rule-no-unknown': [true, { ignoreAtRules: ['function', 'return'] }]
    }
}

```

> Which version of stylelint are you using?

`9.10.1`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

CLI with `stylelint "**/*.css" --config .stylelintrc.js`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

Yes, it's related to css module pre-sudo function like `:global()`

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged:

```shell
test.css
4:1  ✖  Expected selector ".componentName :global(.ant-menu)" to come before selector ".componentName :global(.ant-modal) :global(.ant-btn)"   no-descending-specificity
```

**Comment by jeddy3** [comment]
@vegawong Thanks for the report and for using the template.

It looks like the rule struggles with [_functional pseudo-classes_](https://www.w3.org/TR/selectors-4/#functional-pseudo-class).

I can reproduce this with:

```
:fsc1() :fsc2() {}
:fsc3() {}
```

As the issue has been labelled "help wanted", please consider [contributing a fix](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#fixing-a-bug-in-an-existing-rule).

**Comment by vegawong** [comment]
@jeddy3 , Thanks for reply! I am tring to understand how to calculate the specificity by stylelint. I found that the specificity calculate result was not equal to the `vscode` tips.

**Comment by vegawong** [comment]
Hey guys! Maybe i had found the problem. 
https://github.com/stylelint/stylelint/blob/b6f3057909270a07fc470508a3d0f2c8fefa38b6/lib/rules/no-descending-specificity/index.js#L75
In this line, the `:global` was reduced as a `presudo selector` that all the last node use `:global` as the same `comparisonContext`. Is it right?
`:global` and `:local` are the predefine words in css-module. I suggest that add the word into `keywordSets.pseudoElements` and make it  independent to deal with `comparisonContext`.

**Comment by yoyo837** [comment]
```less
@import '../../variable.less';

@imgSize: 32px;
@gutter: (@menu-collapsed-width - @imgSize) / 2;

.sider,
.logo,
.logo h1 {
  transition: all 0.2s;
}

.sider {
  padding-top: @layout-header-height;
  position: fixed;
  height: 100%;
  left: 0;
  top: 0;
  :global(.ant-layout-sider-children) {
    padding-right: 20px;
    width: calc(~'100% + 20px');
    height: calc(~'100vh - @{layout-header-height}');
    overflow-x: hidden;
    overflow-y: auto;
    -ms-overflow-style: none; // overflow: -moz-scrollbars-none;
    &::-webkit-scrollbar {
      display: none;
    }
  }
}

.logo {
  position: fixed;
  top: 0;
  left: 0;
  height: @layout-header-height;
  line-height: @layout-header-height;
  padding-left: @gutter;
  background: @header-background;
  width: 200px;
  color: @header-color;
  overflow: hidden;
  z-index: 1;
  img,
  h1 {
    display: inline-block;
    vertical-align: middle;
  }
  img {
    height: @imgSize;
  }
  h1 {
    color: #fff;
    font-size: 20px;
    font-family: Avenir, 'Helvetica Neue', Arial, Helvetica, sans-serif;
    font-weight: 600;
    margin: 0 0 0 @gutter;
  }
}

.siderCollapsed {
  .logo {
    padding-left: 24px;
    padding-right: 24px;
    width: auto;
    h1 {
      display: none;
    }
  }
 // here
 // `Expected selector ".siderCollapsed :global(.ant-menu-item-group-title)" to come before selector ".sider :global(.ant-layout-sider-children)::-webkit-scrollbar" (no-descending-specificity)stylelint(no-descending-specificity)`
  :global(.ant-menu-item-group-title) {
    text-align: center;
  }
}

```

**Comment by hazxy** [comment]
same problem

**Comment by ybiquitous** [comment]
The `no-descending-specificity` rule's implementation has now changed and improved by using heavily the [`@csstools/selector-specificity`](https://www.npmjs.com/package/@csstools/selector-specificity) package. See #6131 and the following code for details.

https://github.com/stylelint/stylelint/blob/592b149a6225559cccada30d313814910439d170/lib/rules/no-descending-specificity/index.js#L4

The false positive reported by the issue appears still on the [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0kC2AHCA7GWAuAcgIaowAE8A5gDYQBGR1AFEkfgLSrSMCUFN9RizZ52dPFj7AAOljLyykWgCd4ZAMQAzbQG5ZAX1koIGbLkIlyVWg2asOpLAFcpshYogq1W3QZAAaEABnAE98IgAPBHAgoIDwbE0AS0pomTkyaRBlJ2oYIKy1dPcsrAh2WCCwXCgkrEp2IPQYMCTk1rwQwrI8HJg3MkMsfRB9IA), but I doubt if it's a false positive. 🤔 
Because it seems `@csstools/selector-specificity` properly calculates specificities as below:

```js
const parser = require('postcss-selector-parser');
const { selectorSpecificity } = require('@csstools/selector-specificity');

console.log(selectorSpecificity(parser().astSync('.componentName :global(.ant-modal) :global(.ant-btn)')));
//=> { a: 0, b: 3, c: 0 }

console.log(selectorSpecificity(parser().astSync('.componentName :global(.ant-menu)')));
//=> { a: 0, b: 2, c: 0 }
```

See also https://github.com/csstools/postcss-plugins/blob/28865addbb350c784ade60faf5752ca31add1a93/packages/selector-specificity/src/index.ts#L103-L115

What do you think?

**Comment by jeddy3** [comment]
> The false positive reported by the issue appears still on the [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0kC2AHCA7GWAuAcgIaowAE8A5gDYQBGR1AFEkfgLSrSMCUFN9RizZ52dPFj7AAOljLyykWgCd4ZAMQAzbQG5ZAX1koIGbLkIlyVWg2asOpLAFcpshYogq1W3QZAAaEABnAE98IgAPBHAgoIDwbE0AS0pomTkyaRBlJ2oYIKy1dPcsrAh2WCCwXCgkrEp2IPQYMCTk1rwQwrI8HJg3MkMsfRB9IA), but I doubt if it's a false positive. 🤔

I agree. Let's close.

**Comment by richex-cn** [comment]
Same error for me, How should I fix it ?

**Comment by ybiquitous** [comment]
@richex-cn Could you reproduce an error on the [demo](https://stylelint.io/demo) and paste the demo URL here, please?

**Comment by richex-cn** [comment]
@ybiquitous Sure, It's [demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0YCGAnKACYAdAdllkgM4DuAlgA4w4FFHwDmANhAEaosAUplN6ALQkWFWINRgALhQBuMAJQAaeg2ZtOPPtRhCRYmIKgBXKqLRTDkmfIV1CDIlPSp8JAGYR0AW3hYSaCww3ACMCgDcqkQAvqqx+PEESGyoUBT4TPaMrBxcvLKugilpGQD6pSRU6fi6dngORJRQUgAWfgAMSAAc6DDekQ7x0SBKICQAnvhSqAAeCOAkJCPgEPjuFEzz9US4IOjGQSS7ftsMu-gQRjABMPglTMI0YBTrz1Ljx1jOxjBxBMPRIA).

**Comment by ybiquitous** [comment]
@richex-cn Thanks. I'll look into it. 👍🏼

**Comment by milkyfuckinway** [comment]
Hey, can we reopen this issue? https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0AjArgFwxAdgAmAB188BDAG3PjzRwGcYMBuY4vPAMngHNyIUKACiQAnGFACUBNuzyQ+I6iO4pBAJgCsGvAAZdEljhnsuACwgA3GCOklZciAqUr1WvAEYdegOwaDxvABfVjtOeAAzCDA0Olt7dhQIAA8AWjpTUigIAHdqPXy8NQAHJLxlVU1tTx8-Q3tgkgaZLl5+IVRyNBgpIlD5CEUylw0Cyv8jULNLazj4-sHywRGPL0KtcfqQ+y5I6Nje+LxE1PTMnLzdS+LSxeXqtdqAhvYGwJAAGhBwgEtyGAA5UgAWzgiBgSWBRT+SDAdDoH3AuB+3AQIAOeEIIBEaD+dEx1HR7ExOAgKVgdDAMBwUG+OG4aSKMDA3x+zIwAE98XgMNiYDJXgjIDhkQAxAZA0gYVEAKzouARsCK8MQ6MxdA5f3ItIwXMx5ElMHVmPeMjVGpgWpwGBSQuRaQwpGppBEUF1IH1GENOpAxDegSAA
